# I. Data description
[Mô tả chi tiết](https://docs.google.com/document/d/11S0v9u_XCf-xoS--OKHtF3vom9EuO5LiUBLf1RQy7fw/edit#)

## 1. Data ##
___Training Data:___   Dữ liệu động cơ chạy đến lỗi.
[download training data](https://drive.google.com/drive/u/1/folders/13I9zwIeHENIJSl6SUJChl3sPh1fTgg9A)  
___Test Data:___ Dữ liệu vận hành động cơ mà không có sự kiện hỏng hóc được ghi lại.
[download test data](https://drive.google.com/drive/u/1/folders/13I9zwIeHENIJSl6SUJChl3sPh1fTgg9A)  
___Ground Truth Data:___ Dữ liệu thật: Dữ liệu sự thật cơ bản: Các chu kỳ thực sự còn lại cho mỗi động cơ trong test data.
[download truth data](https://drive.google.com/drive/u/1/folders/13I9zwIeHENIJSl6SUJChl3sPh1fTgg9A)  

## 2. Note (Data Columns)

- col_names = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 's1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 's18', 's19', 's20', 's21']
+ id: là ID động cơ, nằm trong khoảng từ 1 đến 100
+ cycle: cho mỗi chuỗi động cơ, bắt đầu từ 1 đến số chu kỳ đã xảy ra lỗi (test data)
+ setting1 to setting3: cài đặt hoạt động động cơ
+ s1 to s21: các phép đo cảm biến


#  II. Task to do

## 1. Import libraries and packages

In [45]:
import numpy as np
import pandas as pd
import os

## 2. Load Training Data

### 2.1 Load training data

In [46]:
# Loading training data
df_train_raw = pd.read_csv('./data/data_train.txt', sep=' ', header=None)
df_train_raw.head()

0   1       2       3      4       5       6        7        8      9   \
0   1   1 -0.0007 -0.0004  100.0  518.67  641.82  1589.70  1400.60  14.62   
1   1   2  0.0019 -0.0003  100.0  518.67  642.15  1591.82  1403.14  14.62   
2   1   3 -0.0043  0.0003  100.0  518.67  642.35  1587.99  1404.20  14.62   
3   1   4  0.0007  0.0000  100.0  518.67  642.35  1582.79  1401.87  14.62   
4   1   5 -0.0019 -0.0002  100.0  518.67  642.37  1582.85  1406.22  14.62   

   ...       18      19    20   21    22     23     24       25  26  27  
0  ...  8138.62  8.4195  0.03  392  2388  100.0  39.06  23.4190 NaN NaN  
1  ...  8131.49  8.4318  0.03  392  2388  100.0  39.00  23.4236 NaN NaN  
2  ...  8133.23  8.4178  0.03  390  2388  100.0  38.95  23.3442 NaN NaN  
3  ...  8133.83  8.3682  0.03  392  2388  100.0  38.88  23.3739 NaN NaN  
4  ...  8133.80  8.4294  0.03  393  2388  100.0  38.90  23.4044 NaN NaN  

[5 rows x 28 columns]

### 2.2 Gán tên cột theo Data description

In [47]:
# Gán tên cột
df_train_raw = df_train_raw[[s for s in range(0,26)]]
df_train_raw.columns = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 
                        's1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9', 
                        's10','s11', 's12', 's13', 's14', 's15', 's16', 's17', 
                        's18', 's19', 's20', 's21']
df_train_raw.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...     s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62  ...  521.66  2388.02  8138.62  8.4195  0.03  392  2388  100.0   
1  14.62  ...  522.28  2388.07  8131.49  8.4318  0.03  392  2388  100.0   
2  14.62  ...  522.42  2388.03  8133.23  8.4178  0.03  390  2388  100.0   
3  14.62  ...  522.86  2388.08  8133.83  8.3682  0.03  392  2388  100.0   
4  14.62  ...  522.19  2388.04  8133.80  8.4294  0.03  393  2388  100.0   

     s20      s21  
0  39.06  23.4190  
1  39.00  23.4236  
2  38.95  23.3442  
3  38.88  23.3739  
4  38.90  23.4044  

[5 rows x 26 columns]

### 2.3 Thông tin từng cột có trong dataframe

In [48]:
# Hiển thị thông tin từng cột
# Kiểm tra kiểu dữ liệu của từng cột
df_train_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20631 entries, 0 to 20630
Data columns (total 26 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        20631 non-null  int64  
 1   cycle     20631 non-null  int64  
 2   setting1  20631 non-null  float64
 3   setting2  20631 non-null  float64
 4   setting3  20631 non-null  float64
 5   s1        20631 non-null  float64
 6   s2        20631 non-null  float64
 7   s3        20631 non-null  float64
 8   s4        20631 non-null  float64
 9   s5        20631 non-null  float64
 10  s6        20631 non-null  float64
 11  s7        20631 non-null  float64
 12  s8        20631 non-null  float64
 13  s9        20631 non-null  float64
 14  s10       20631 non-null  float64
 15  s11       20631 non-null  float64
 16  s12       20631 non-null  float64
 17  s13       20631 non-null  float64
 18  s14       20631 non-null  float64
 19  s15       20631 non-null  float64
 20  s16       20631 non-null  fl

__Tất cả dữ liệu đều là số__

### 2.4 Kiểm tra NaN values

In [49]:
# Kiểm tra missing data
df_train_raw.isnull().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

__Không có giá trị nào bị thiếu (missing). Dữ liệu sạch.__

### 2.5 Hiển thị id có trong Dataframe

In [50]:
# Check xem id có giống data description mô tả.
df_train_raw.id.unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100], dtype=int64)

__ID: là ID động cơ nằm trong khoảng từ 1 đến 100 đúng như Data description.__

### 2.6 Cycle - Chu kỳ

In [53]:
# Check cycle - chu kỳ dộng cơ có giống data description
df_train_raw.cycle.unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

__cycle: cho mỗi chuỗi động cơ, bắt đầu từ 1 đến số chu kỳ đã xảy ra lỗi (chỉ dữ liệu đào tạo)__

In [54]:
# Nhận thấy có nhiều sự lập lại ở ID
# Kiểm tra xem có bao nhiêu id == 1
df_train_raw[df_train_raw['id'] == 1]

id  cycle  setting1  setting2  setting3      s1      s2       s3  \
0     1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70   
1     1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82   
2     1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99   
3     1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79   
4     1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85   
..   ..    ...       ...       ...       ...     ...     ...      ...   
187   1    188   -0.0067    0.0003     100.0  518.67  643.75  1602.38   
188   1    189   -0.0006    0.0002     100.0  518.67  644.18  1596.17   
189   1    190   -0.0027    0.0001     100.0  518.67  643.64  1599.22   
190   1    191   -0.0000   -0.0004     100.0  518.67  643.34  1602.36   
191   1    192    0.0009   -0.0000     100.0  518.67  643.54  1601.41   

          s4     s5  ...     s12      s13      s14     s15   s16  s17   s18  \
0    1400.60  14.62  ...  521.66  2388.02  8138.62  8.4195  0.03  392  2388   
1    1403.14  14.62  ...  522.28  2388.07  8131.49  8.4318  0.03  392  2388   
2    1404.20  14.62  ...  522.42  2388.03  8133.23  8.4178  0.03  390  2388   
3    1401.87  14.62  ...  522.86  2388.08  8133.83  8.3682  0.03  392  2388   
4    1406.22  14.62  ...  522.19  2388.04  8133.80  8.4294  0.03  393  2388   
..       ...    ...  ...     ...      ...      ...     ...   ...  ...   ...   
187  1422.78  14.62  ...  519.79  2388.23  8117.69  8.5207  0.03  396  2388   
188  1428.01  14.62  ...  519.58  2388.33  8117.51  8.5183  0.03  395  2388   
189  1425.95  14.62  ...  520.04  2388.35  8112.58  8.5223  0.03  398  2388   
190  1425.77  14.62  ...  519.57  2388.30  8114.61  8.5174  0.03  394  2388   
191  1427.20  14.62  ...  520.08  2388.32  8110.93  8.5113  0.03  396  2388   

       s19    s20      s21  
0    100.0  39.06  23.4190  
1    100.0  39.00  23.4236  
2    100.0  38.95  23.3442  
3    100.0  38.88  23.3739  
4    100.0  38.90  23.4044  
..     ...    ...      ...  
187  100.0  38.51  22.9588  
188  100.0  38.48  23.1127  
189  100.0  38.49  23.0675  
190  100.0  38.45  23.1295  
191  100.0  38.48  22.9649  

[192 rows x 26 columns]

__Có tất cả 192 ID bằng 1__

In [55]:
# Tính max cycle của từng id
max_cycles_df = df_train_raw.groupby(['id'], sort=False)['cycle'].max().reset_index().rename(columns={'cycle':'MaxCycleID'})
max_cycles_df.head()

id  MaxCycleID
0   1         192
1   2         287
2   3         179
3   4         189
4   5         269

In [56]:
# Tính tuổi thọ tối đa của động cơ
df_train_raw[["id","cycle"]].groupby("id").max().sort_values(by = ["cycle"], ascending= False).head(5)

cycle
id       
69    362
92    341
96    336
67    313
83    293

### 2.7 Tính Remaining Useful Life (RUL) - Tuổi thọ hữu ít còn lại

In [57]:
# Tính Remaining Useful Life (RUL) và merge vào Dataframe
rul_df = pd.merge(df_train_raw, max_cycles_df, how='inner', on='id')
rul_df['RUL'] = rul_df['MaxCycleID'] - rul_df['cycle']
rul_df

id  cycle  setting1  setting2  setting3      s1      s2       s3  \
0        1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70   
1        1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82   
2        1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99   
3        1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79   
4        1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85   
...    ...    ...       ...       ...       ...     ...     ...      ...   
20626  100    196   -0.0004   -0.0003     100.0  518.67  643.49  1597.98   
20627  100    197   -0.0016   -0.0005     100.0  518.67  643.54  1604.50   
20628  100    198    0.0004    0.0000     100.0  518.67  643.42  1602.46   
20629  100    199   -0.0011    0.0003     100.0  518.67  643.23  1605.26   
20630  100    200   -0.0032   -0.0005     100.0  518.67  643.85  1600.38   

            s4     s5  ...      s14     s15   s16  s17   s18    s19    s20  \
0      1400.60  14.62  ...  8138.62  8.4195  0.03  392  2388  100.0  39.06   
1      1403.14  14.62  ...  8131.49  8.4318  0.03  392  2388  100.0  39.00   
2      1404.20  14.62  ...  8133.23  8.4178  0.03  390  2388  100.0  38.95   
3      1401.87  14.62  ...  8133.83  8.3682  0.03  392  2388  100.0  38.88   
4      1406.22  14.62  ...  8133.80  8.4294  0.03  393  2388  100.0  38.90   
...        ...    ...  ...      ...     ...   ...  ...   ...    ...    ...   
20626  1428.63  14.62  ...  8137.60  8.4956  0.03  397  2388  100.0  38.49   
20627  1433.58  14.62  ...  8136.50  8.5139  0.03  395  2388  100.0  38.30   
20628  1428.18  14.62  ...  8141.05  8.5646  0.03  398  2388  100.0  38.44   
20629  1426.53  14.62  ...  8139.29  8.5389  0.03  395  2388  100.0  38.29   
20630  1432.14  14.62  ...  8137.33  8.5036  0.03  396  2388  100.0  38.37   

           s21  MaxCycleID  RUL  
0      23.4190         192  191  
1      23.4236         192  190  
2      23.3442         192  189  
3      23.3739         192  188  
4      23.4044         192  187  
...        ...         ...  ...  
20626  22.9735         200    4  
20627  23.1594         200    3  
20628  22.9333         200    2  
20629  23.0640         200    1  
20630  23.0522         200    0  

[20631 rows x 28 columns]

In [58]:
rul_df['RUL'].head()

0    191
1    190
2    189
3    188
4    187
Name: RUL, dtype: int64

In [59]:
# Hiển thị cycle và RUL
rul_df[['cycle', 'RUL']]

cycle  RUL
0          1  191
1          2  190
2          3  189
3          4  188
4          5  187
...      ...  ...
20626    196    4
20627    197    3
20628    198    2
20629    199    1
20630    200    0

[20631 rows x 2 columns]

In [60]:
# Kiểm và hiển thị thông tin vừa thêm vào Dataframe
rul_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20631 entries, 0 to 20630
Data columns (total 28 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          20631 non-null  int64  
 1   cycle       20631 non-null  int64  
 2   setting1    20631 non-null  float64
 3   setting2    20631 non-null  float64
 4   setting3    20631 non-null  float64
 5   s1          20631 non-null  float64
 6   s2          20631 non-null  float64
 7   s3          20631 non-null  float64
 8   s4          20631 non-null  float64
 9   s5          20631 non-null  float64
 10  s6          20631 non-null  float64
 11  s7          20631 non-null  float64
 12  s8          20631 non-null  float64
 13  s9          20631 non-null  float64
 14  s10         20631 non-null  float64
 15  s11         20631 non-null  float64
 16  s12         20631 non-null  float64
 17  s13         20631 non-null  float64
 18  s14         20631 non-null  float64
 19  s15         20631 non-nul

### 2.8 Hiển thị chỉ số cần thiết

In [61]:
# Chỉ sổ cần thiết
df_train_raw.describe()

id         cycle      setting1      setting2  setting3  \
count  20631.000000  20631.000000  20631.000000  20631.000000   20631.0   
mean      51.506568    108.807862     -0.000009      0.000002     100.0   
std       29.227633     68.880990      0.002187      0.000293       0.0   
min        1.000000      1.000000     -0.008700     -0.000600     100.0   
25%       26.000000     52.000000     -0.001500     -0.000200     100.0   
50%       52.000000    104.000000      0.000000      0.000000     100.0   
75%       77.000000    156.000000      0.001500      0.000300     100.0   
max      100.000000    362.000000      0.008700      0.000600     100.0   

                 s1            s2            s3            s4            s5  \
count  2.063100e+04  20631.000000  20631.000000  20631.000000  2.063100e+04   
mean   5.186700e+02    642.680934   1590.523119   1408.933782  1.462000e+01   
std    6.537152e-11      0.500053      6.131150      9.000605  3.394700e-12   
min    5.186700e+02    641.210000   1571.040000   1382.250000  1.462000e+01   
25%    5.186700e+02    642.325000   1586.260000   1402.360000  1.462000e+01   
50%    5.186700e+02    642.640000   1590.100000   1408.040000  1.462000e+01   
75%    5.186700e+02    643.000000   1594.380000   1414.555000  1.462000e+01   
max    5.186700e+02    644.530000   1616.910000   1441.490000  1.462000e+01   

       ...           s12           s13           s14           s15  \
count  ...  20631.000000  20631.000000  20631.000000  20631.000000   
mean   ...    521.413470   2388.096152   8143.752722      8.442146   
std    ...      0.737553      0.071919     19.076176      0.037505   
min    ...    518.690000   2387.880000   8099.940000      8.324900   
25%    ...    520.960000   2388.040000   8133.245000      8.414900   
50%    ...    521.480000   2388.090000   8140.540000      8.438900   
75%    ...    521.950000   2388.140000   8148.310000      8.465600   
max    ...    523.380000   2388.560000   8293.720000      8.584800   

                s16           s17      s18      s19           s20  \
count  2.063100e+04  20631.000000  20631.0  20631.0  20631.000000   
mean   3.000000e-02    393.210654   2388.0    100.0     38.816271   
std    1.556432e-14      1.548763      0.0      0.0      0.180746   
min    3.000000e-02    388.000000   2388.0    100.0     38.140000   
25%    3.000000e-02    392.000000   2388.0    100.0     38.700000   
50%    3.000000e-02    393.000000   2388.0    100.0     38.830000   
75%    3.000000e-02    394.000000   2388.0    100.0     38.950000   
max    3.000000e-02    400.000000   2388.0    100.0     39.430000   

                s21  
count  20631.000000  
mean      23.289705  
std        0.108251  
min       22.894200  
25%       23.221800  
50%       23.297900  
75%       23.366800  
max       23.618400  

[8 rows x 26 columns]

* count: Đếm số quan sát.
* maen: Trung bình các giá trị
* std: Standard deviation - Độ lệch chuẩn của các quan sát.
* min: Min trong đối tượng.
* max: Max các giá trị trong đối tượng.

__Có 100 động cơ. mỗi động cơ có từ 1 đến 362 vòng (trung bình 108 vòng cho mỗi động cơ).__

## 3.  Load Test Data

### 3.1 Load test data

In [62]:
#load test data
df_test_raw = pd.read_csv('./data/data_test.txt', sep = ' ', header=None)
df_test_raw

0    1       2       3      4       5       6        7        8   \
0        1    1  0.0023  0.0003  100.0  518.67  643.02  1585.29  1398.21   
1        1    2 -0.0027 -0.0003  100.0  518.67  641.71  1588.45  1395.42   
2        1    3  0.0003  0.0001  100.0  518.67  642.46  1586.94  1401.34   
3        1    4  0.0042  0.0000  100.0  518.67  642.44  1584.12  1406.42   
4        1    5  0.0014  0.0000  100.0  518.67  642.51  1587.19  1401.92   
...    ...  ...     ...     ...    ...     ...     ...      ...      ...   
13091  100  194  0.0049  0.0000  100.0  518.67  643.24  1599.45  1415.79   
13092  100  195 -0.0011 -0.0001  100.0  518.67  643.22  1595.69  1422.05   
13093  100  196 -0.0006 -0.0003  100.0  518.67  643.44  1593.15  1406.82   
13094  100  197 -0.0038  0.0001  100.0  518.67  643.26  1594.99  1419.36   
13095  100  198  0.0013  0.0003  100.0  518.67  642.95  1601.62  1424.99   

          9   ...       18      19    20   21    22     23     24       25  \
0      14.62  ...  8125.55  8.4052  0.03  392  2388  100.0  38.86  23.3735   
1      14.62  ...  8139.62  8.3803  0.03  393  2388  100.0  39.02  23.3916   
2      14.62  ...  8130.10  8.4441  0.03  393  2388  100.0  39.08  23.4166   
3      14.62  ...  8132.90  8.3917  0.03  391  2388  100.0  39.00  23.3737   
4      14.62  ...  8129.54  8.4031  0.03  390  2388  100.0  38.99  23.4130   
...      ...  ...      ...     ...   ...  ...   ...    ...    ...      ...   
13091  14.62  ...  8213.28  8.4715  0.03  394  2388  100.0  38.65  23.1974   
13092  14.62  ...  8210.85  8.4512  0.03  395  2388  100.0  38.57  23.2771   
13093  14.62  ...  8217.24  8.4569  0.03  395  2388  100.0  38.62  23.2051   
13094  14.62  ...  8220.48  8.4711  0.03  395  2388  100.0  38.66  23.2699   
13095  14.62  ...  8214.64  8.4903  0.03  396  2388  100.0  38.70  23.1855   

       26  27  
0     NaN NaN  
1     NaN NaN  
2     NaN NaN  
3     NaN NaN  
4     NaN NaN  
...    ..  ..  
13091 NaN NaN  
13092 NaN NaN  
13093 NaN NaN  
13094 NaN NaN  
13095 NaN NaN  

[13096 rows x 28 columns]

### 3.2 Gán tên cột theo Data description

In [63]:
# Gán tên cột
df_test_raw = df_test_raw[[s for s in range(0,26)]]
df_test_raw.columns = ['id', 'cycle', 'setting1', 'setting2', 'setting3', 
                   's1', 's2', 's3', 's4', 's5', 's6', 's7', 's8', 's9',
                   's10', 's11', 's12', 's13', 's14', 's15', 's16', 's17', 
                   's18', 's19', 's20', 's21']
df_test_raw.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29  1398.21   
1   1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45  1395.42   
2   1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94  1401.34   
3   1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12  1406.42   
4   1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19  1401.92   

      s5  ...     s12      s13      s14     s15   s16  s17   s18    s19  \
0  14.62  ...  521.72  2388.03  8125.55  8.4052  0.03  392  2388  100.0   
1  14.62  ...  522.16  2388.06  8139.62  8.3803  0.03  393  2388  100.0   
2  14.62  ...  521.97  2388.03  8130.10  8.4441  0.03  393  2388  100.0   
3  14.62  ...  521.38  2388.05  8132.90  8.3917  0.03  391  2388  100.0   
4  14.62  ...  522.15  2388.03  8129.54  8.4031  0.03  390  2388  100.0   

     s20      s21  
0  38.86  23.3735  
1  39.02  23.3916  
2  39.08  23.4166  
3  39.00  23.3737  
4  38.99  23.4130  

[5 rows x 26 columns]

### 3.3 Hiển thị chỉ số cần thiết trong Test Data

In [64]:
# Chỉ số cần thiết trong test data
df_test_raw.describe()

id         cycle      setting1      setting2  setting3  \
count  13096.000000  13096.000000  13096.000000  13096.000000   13096.0   
mean      51.543907     76.836515     -0.000011      0.000004     100.0   
std       28.289423     53.057749      0.002203      0.000294       0.0   
min        1.000000      1.000000     -0.008200     -0.000600     100.0   
25%       28.000000     33.000000     -0.001500     -0.000200     100.0   
50%       52.000000     69.000000      0.000000      0.000000     100.0   
75%       76.000000    113.000000      0.001500      0.000300     100.0   
max      100.000000    303.000000      0.007800      0.000700     100.0   

                 s1            s2            s3            s4            s5  \
count  1.309600e+04  13096.000000  13096.000000  13096.000000  1.309600e+04   
mean   5.186700e+02    642.475088   1588.099204   1404.735362  1.462000e+01   
std    6.014263e-11      0.400899      5.003274      6.688309  2.671743e-12   
min    5.186700e+02    641.130000   1569.040000   1384.390000  1.462000e+01   
25%    5.186700e+02    642.197500   1584.600000   1399.950000  1.462000e+01   
50%    5.186700e+02    642.460000   1587.990000   1404.440000  1.462000e+01   
75%    5.186700e+02    642.740000   1591.362500   1409.050000  1.462000e+01   
max    5.186700e+02    644.300000   1607.550000   1433.360000  1.462000e+01   

       ...           s12           s13           s14           s15  \
count  ...  13096.000000  13096.000000  13096.000000  13096.000000   
mean   ...    521.747724   2388.071025   8138.947817      8.425844   
std    ...      0.559627      0.056934     10.188605      0.029009   
min    ...    519.380000   2387.890000   8108.500000      8.332800   
25%    ...    521.380000   2388.030000   8132.310000      8.405600   
50%    ...    521.780000   2388.070000   8138.390000      8.424900   
75%    ...    522.150000   2388.110000   8144.360000      8.444300   
max    ...    523.760000   2388.320000   8220.480000      8.541400   

                s16           s17      s18      s19           s20  \
count  1.309600e+04  13096.000000  13096.0  13096.0  13096.000000   
mean   3.000000e-02    392.571625   2388.0    100.0     38.892502   
std    8.823140e-15      1.233577      0.0      0.0      0.141681   
min    3.000000e-02    389.000000   2388.0    100.0     38.310000   
25%    3.000000e-02    392.000000   2388.0    100.0     38.800000   
50%    3.000000e-02    393.000000   2388.0    100.0     38.900000   
75%    3.000000e-02    393.000000   2388.0    100.0     38.990000   
max    3.000000e-02    397.000000   2388.0    100.0     39.410000   

                s21  
count  13096.000000  
mean      23.335743  
std        0.084120  
min       22.935400  
25%       23.281600  
50%       23.339200  
75%       23.393600  
max       23.641900  

[8 rows x 26 columns]

* count: Đếm số quan sát.
* maen: Trung bình các giá trị
* std: Standard deviation - Độ lệch chuẩn của các quan sát.
* min: Min trong đối tượng.
* max: Max các giá trị trong đối tượng.

In [67]:
df_test_raw.cycle.unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

__Có 100 động cơ, mỗi động cơ có từ 1 đến 303 chu kỳ (trung bình 76 chu kỳ cho mỗi động cơ).__

## 4. Load Truth Data

__Note: TTF - Time-To-Failure: Thời gian không hỏng hóc__

### 4.1 Load truth data - 'ttf' thực tế cho test data

In [68]:
# load truth data 
df_truth = pd.read_csv('./data/data_truth.txt', sep = ' ', header=None)
df_truth

0   1
0   112 NaN
1    98 NaN
2    69 NaN
3    82 NaN
4    91 NaN
..  ...  ..
95  137 NaN
96   82 NaN
97   59 NaN
98  117 NaN
99   20 NaN

[100 rows x 2 columns]

### 4.2  TTF

In [69]:
# drop cột 1:NaN, đổi tên cộkt 0 = 'ttf' (TTF - Time-To-Failure: Thời gian không hỏng hóc)
df_truth.drop([1], axis=1, inplace=True)
df_truth.columns = ['ttf']
df_truth.head()

ttf
0  112
1   98
2   69
3   82
4   91

### 4.3 Chỉ số cần thiết Truth data

In [70]:
# Chỉ số cần thiết trong truth data
df_truth.describe()

ttf
count  100.00000
mean    75.52000
std     41.76497
min      7.00000
25%     32.75000
50%     86.00000
75%    112.25000
max    145.00000

__Để có được test data có ý nghĩa, hợp nhất truth data (TTF) với chu kỳ cuối cùng cho mỗi động cơ trong test data.__

__Cung cấp 1 bộ test gồm 100 động cơ với dữ liệu TTF của chúng.__ 

__Điều này xảy ra khi tạo nhãn hồi quy và phân loại cho traininng và test.__

## 5. Feature 

### 5.1 Add Features

In [71]:
# Add features làm cho thời gian cảm biến mượt hơn bằng cách thêm rolling mean (Đường trung bình động)
# và standard deviation(độ lệch chuẩn)
def add_features(df_in, rolling_win_size):
    
    ''' Add rolling average and rolling standard deviation cho cảm biến, 
    sử dụng cố định rolling window size.
            
           df_in (dataframe):  inppput của dataframe sẽ được sử lí (training or test). 
           rolling_win_size (int): window size, số chu kỳ áp dụng cho rolling function.
           
           dataframe: contains chứa inpput dataframe thêm vào với rolling mean and std cho cảm biến.
    
    '''
    
    sensor_cols = ['s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11',
                's12','s13','s14','s15','s16','s17','s18','s19','s20','s21']
    
    sensor_av_cols = [nm.replace('s', 'av') for nm in sensor_cols]
    sensor_sd_cols = [nm.replace('s', 'sd') for nm in sensor_cols]
    
    df_out = pd.DataFrame()
    
    ws = rolling_win_size
    
    # Tính rolling stats cho id động cơ
    
    for m_id in pd.unique(df_in.id):
    
        # get subset (tập con) cho cảm biến của động cơ
        df_engine = df_in[df_in['id'] == m_id]
        df_sub = df_engine[sensor_cols]

    
        # get rolling mean cho subset
        av = df_sub.rolling(ws, min_periods=1).mean()
        av.columns = sensor_av_cols
    
        # get the rolling standard deviation cho tập con
        sd = df_sub.rolling(ws, min_periods=1).std().fillna(0)
        sd.columns = sensor_sd_cols
    
        # merge 2 cột tập con mới vào dataframe 
        new_ftrs = pd.concat([df_engine,av,sd], axis=1)
    
        # add thêm hàng cho features mới vào output dataframe
        df_out = pd.concat([df_out,new_ftrs])
        
    return df_out

### 5.2 Add the regression and classification labels training data

In [73]:
def prepare_train_data (df_in, period):
    
    ''' Thêm regression and classification labels cho training data.
            
            Regression label: ttf (time-to-failure) = last cycle - cycle
            Binary classification label: label_bnc = if ttf <= chu kỳ parameter then 1 else 0 (values = 0,1)
        
        Args:    
            df_in (dataframe): The input training data
            period (int): Sổ cycles để phân loại TTF dùng classification labels để phân loại
               
        
        Returns: 
            dataframe: Input dataframe with regression and classification labels đã được thêm.
    '''
    
    # Tạo regression label 
    df_max_cycle = pd.DataFrame(df_in.groupby('id')['cycle'].max())
    df_max_cycle.reset_index(level=0, inplace=True)
    df_max_cycle.columns = ['id', 'last_cycle']
    
    #Thêm  mới cột time-to-failure ttf 
    df_in = pd.merge(df_in, df_max_cycle, on='id')
    df_in['ttf'] = df_in['last_cycle'] - df_in['cycle']
    df_in.drop(['last_cycle'], axis=1, inplace=True)
    
    #Tạo binary classification label
    df_in['label_bnc'] = df_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
    #Tạo multi-class classification label
    df_in['label_mcc'] = df_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)

    return df_in

### 5.3 Add the regression and classification labels test data

In [74]:
def prepare_test_data(df_test_in, df_truth_in, period):
    
    '''Add regression and classification labels cho test data.
    
             Regression label: ttf (time-to-failure) = extract cycle cuối cho động cơ và merge nó với truth data
             Binary classification label: label_bnc = if ttf <= chu kỳ parameter then 1 else 0 (values = 0,1)
             
        Args:
          df_in (dataframe): The input training data
          period (int): Sổ cycles để phân loại TTF dùng classification labels để phân loại
          
        Returns: 
            dataframe: Input dataframe with regression and classification labels đã được thêm.
          
    '''
    
    df_tst_last_cycle = pd.DataFrame(df_test_in.groupby('id')['cycle'].max())
    
    df_tst_last_cycle.reset_index(level=0, inplace=True)
    df_tst_last_cycle.columns = ['id', 'last_cycle']
    
    df_test_in = pd.merge(df_test_in, df_tst_last_cycle, on='id')


    df_test_in = df_test_in[df_test_in['cycle'] == df_test_in['last_cycle']]

    df_test_in.drop(['last_cycle'], axis=1, inplace=True)
    
    df_test_in.reset_index(drop=True, inplace=True)
    
    df_test_in = pd.concat([df_test_in, df_truth], axis=1)
    
    #Tạo binary classification label
    df_test_in['label_bnc'] = df_test_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
     #Tạo multi-class classification label
    df_test_in['label_mcc'] = df_test_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)


    return df_test_in

## 6. Prepare the Training Data

### 6.1 Add  features cho training data

In [75]:
df_train_fx = add_features(df_train_raw, 5)
df_train_fx.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...      sd12      sd13      sd14      sd15  sd16      sd17  sd18  \
0  14.62  ...  0.000000  0.000000  0.000000  0.000000   0.0  0.000000   0.0   
1  14.62  ...  0.438406  0.035355  5.041671  0.008697   0.0  0.000000   0.0   
2  14.62  ...  0.404475  0.026458  3.717450  0.007640   0.0  1.154701   0.0   
3  14.62  ...  0.495950  0.029439  3.050906  0.028117   0.0  1.000000   0.0   
4  14.62  ...  0.432574  0.025884  2.651326  0.025953   0.0  1.095445   0.0   

   sd19      sd20      sd21  
0   0.0  0.000000  0.000000  
1   0.0  0.042426  0.003253  
2   0.0  0.055076  0.044573  
3   0.0  0.076322  0.037977  
4   0.0  0.073621  0.033498  

[5 rows x 68 columns]

### 6.2 Add labels

In [76]:
#add labels to training data thời gian sử dụng của 30 cycles cho classification

df_train = prepare_train_data(df_train_fx, 5)
df_train.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1   -0.0007   -0.0004     100.0  518.67  641.82  1589.70  1400.60   
1   1      2    0.0019   -0.0003     100.0  518.67  642.15  1591.82  1403.14   
2   1      3   -0.0043    0.0003     100.0  518.67  642.35  1587.99  1404.20   
3   1      4    0.0007    0.0000     100.0  518.67  642.35  1582.79  1401.87   
4   1      5   -0.0019   -0.0002     100.0  518.67  642.37  1582.85  1406.22   

      s5  ...      sd15  sd16      sd17  sd18  sd19      sd20      sd21  ttf  \
0  14.62  ...  0.000000   0.0  0.000000   0.0   0.0  0.000000  0.000000  191   
1  14.62  ...  0.008697   0.0  0.000000   0.0   0.0  0.042426  0.003253  190   
2  14.62  ...  0.007640   0.0  1.154701   0.0   0.0  0.055076  0.044573  189   
3  14.62  ...  0.028117   0.0  1.000000   0.0   0.0  0.076322  0.037977  188   
4  14.62  ...  0.025953   0.0  1.095445   0.0   0.0  0.073621  0.033498  187   

   label_bnc  label_mcc  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0  

[5 rows x 71 columns]

In [77]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20631 entries, 0 to 20630
Data columns (total 71 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         20631 non-null  int64  
 1   cycle      20631 non-null  int64  
 2   setting1   20631 non-null  float64
 3   setting2   20631 non-null  float64
 4   setting3   20631 non-null  float64
 5   s1         20631 non-null  float64
 6   s2         20631 non-null  float64
 7   s3         20631 non-null  float64
 8   s4         20631 non-null  float64
 9   s5         20631 non-null  float64
 10  s6         20631 non-null  float64
 11  s7         20631 non-null  float64
 12  s8         20631 non-null  float64
 13  s9         20631 non-null  float64
 14  s10        20631 non-null  float64
 15  s11        20631 non-null  float64
 16  s12        20631 non-null  float64
 17  s13        20631 non-null  float64
 18  s14        20631 non-null  float64
 19  s15        20631 non-null  float64
 20  s16   

__Rolling average, rolling standard deviation, regression labels, and classification labels được thêm vào training data__

### 6.3 Save the training data to csv file

In [78]:
df_train.to_csv('data/train.csv', index=False)

## 7. Prepare the Test Data

### 7.1 Add  features to test data

In [79]:
df_test_fx = add_features(df_test_raw, 5)
df_test_fx.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1      1    0.0023    0.0003     100.0  518.67  643.02  1585.29  1398.21   
1   1      2   -0.0027   -0.0003     100.0  518.67  641.71  1588.45  1395.42   
2   1      3    0.0003    0.0001     100.0  518.67  642.46  1586.94  1401.34   
3   1      4    0.0042    0.0000     100.0  518.67  642.44  1584.12  1406.42   
4   1      5    0.0014    0.0000     100.0  518.67  642.51  1587.19  1401.92   

      s5  ...      sd12      sd13      sd14      sd15  sd16      sd17  sd18  \
0  14.62  ...  0.000000  0.000000  0.000000  0.000000   0.0  0.000000   0.0   
1  14.62  ...  0.311127  0.021213  9.948992  0.017607   0.0  0.707107   0.0   
2  14.62  ...  0.220681  0.017321  7.179807  0.032155   0.0  0.577350   0.0   
3  14.62  ...  0.337182  0.015000  5.890096  0.027781   0.0  0.957427   0.0   
4  14.62  ...  0.329742  0.014142  5.222300  0.024080   0.0  1.303840   0.0   

   sd19      sd20      sd21  
0   0.0  0.000000  0.000000  
1   0.0  0.113137  0.012799  
2   0.0  0.113725  0.021642  
3   0.0  0.093095  0.020353  
4   0.0  0.080623  0.020672  

[5 rows x 68 columns]

### 7.2 Add labels to test data

In [80]:
#add labels cho test data thời gian sử dụng của 30 cycles cho classification
df_test = prepare_test_data(df_test_fx, df_truth, 30)
df_test.head()

id  cycle  setting1  setting2  setting3      s1      s2       s3       s4  \
0   1     31   -0.0006    0.0004     100.0  518.67  642.58  1581.22  1398.91   
1   2     49    0.0018   -0.0001     100.0  518.67  642.55  1586.59  1410.83   
2   3    126   -0.0016    0.0004     100.0  518.67  642.88  1589.75  1418.89   
3   4    106    0.0012    0.0004     100.0  518.67  642.78  1594.53  1406.88   
4   5     98   -0.0013   -0.0004     100.0  518.67  642.27  1589.94  1419.36   

      s5  ...      sd15  sd16      sd17  sd18  sd19      sd20      sd21  ttf  \
0  14.62  ...  0.021174   0.0  0.707107   0.0   0.0  0.124016  0.025037  112   
1  14.62  ...  0.011429   0.0  1.095445   0.0   0.0  0.069785  0.027820   98   
2  14.62  ...  0.030924   0.0  0.836660   0.0   0.0  0.143631  0.018966   69   
3  14.62  ...  0.021249   0.0  1.341641   0.0   0.0  0.126372  0.021106   82   
4  14.62  ...  0.009627   0.0  0.547723   0.0   0.0  0.091924  0.101100   91   

   label_bnc  label_mcc  
0          0          0  
1          0          0  
2          0          0  
3          0          0  
4          0          0  

[5 rows x 71 columns]

In [81]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 71 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   id         100 non-null    int64  
 1   cycle      100 non-null    int64  
 2   setting1   100 non-null    float64
 3   setting2   100 non-null    float64
 4   setting3   100 non-null    float64
 5   s1         100 non-null    float64
 6   s2         100 non-null    float64
 7   s3         100 non-null    float64
 8   s4         100 non-null    float64
 9   s5         100 non-null    float64
 10  s6         100 non-null    float64
 11  s7         100 non-null    float64
 12  s8         100 non-null    float64
 13  s9         100 non-null    float64
 14  s10        100 non-null    float64
 15  s11        100 non-null    float64
 16  s12        100 non-null    float64
 17  s13        100 non-null    float64
 18  s14        100 non-null    float64
 19  s15        100 non-null    float64
 20  s16        

### 7.3 Save the test data to csv file

In [83]:
df_test.to_csv('data/test.csv', index=False)

__Rolling average, rolling standard deviation, regression labels, and classification labels được thêm vào test data__